src/EventLoop.php

```diff
@@ -67,6 +71,29 @@ protected function now(): float
         }
     }
 
+    private static function onShutdown(int $invocationCount): void
+    {
+        \gc_collect_cycles();
+
+        $driver = self::getDriver();
+
+        $pending = false;
+        foreach ($driver->getIdentifiers() as $identifier) {
```

Comment:
Shouldn't this also check for pending microtasks/queued callbacks?

Reply:
Yes, you're right. I guess `AbstractDriver::isEmpty()` also isn't doing that and maybe should (though that function is private and due to how the queue is invoked, I'm not sure it ever can have queued microtasks when isEmpty() is called).

Unfortunately there is nothing on `Driver` which will let me query if there are queued microtasks, so that would need to be added.

Reply:
In my code, I just unconditionally do

```
                if (!EventLoop::getDriver()->isRunning()) {
                        $delay = PHP_SAPI === 'cli' ? 5 : 1;
                        $id2 = null;
                        EventLoop::unreference($id1 = EventLoop::delay((float)$delay, static function () use ($delay, &$id2): void {
                                $t = time() - $delay;
                                EventLoop::unreference($id2 = EventLoop::repeat(1.0, static function () use ($t): void {
                                        $t = time() - $t;
                                        Logger::get()->error("Still waiting for shutdown of the event loop for {$t} seconds...");
                                }));
                        }));
                        EventLoop::run();
                        EventLoop::cancel($id1);
                        if ($id2 !== null) {
                                EventLoop::cancel($id2);
                        }
                }
```

Reply:
i.e. an unconditional `run()`